Adding ModernBert support

pappitti opened the pull request:
Adding support for the ModernBERT architecture.  
Because we probably want to be able to deal with non-sentence-transformers models, a new abstraction was added (optional for users). This is inspired by how HF transformers deals with the various use-cases for these models.  

Note 1 : see comments in README (in code mode)
Note 2: some of the XLM-Roberta models I tested didn't work (see comments in README). 

![Frame 33 (1)](https://github.com/user-attachments/assets/867abc02-e886-4662-9ae7-4fc5a1394c32)

Blaizzy commented:
Hi @pappitti 

Great work!

I really love the concept of pipelines.

Could you clean up the code base (i.e, remove all the extra comments and run > pre-commit run --all)

The comments you can add the to the PR description with file names.

Additionally, after looking at the code I'm not 100% sure we need both `ModelForSentenceTransformers` class. We can load the sentence-transformers config to `ModelForSentenceSimilarity` (if it exists).

Blaizzy commented on mlx_embeddings/models/base.py:
```diff
@@ -13,3 +14,59 @@ def from_dict(cls, params):
                 if k in inspect.signature(cls).parameters
             }
         )
+
+def rotate_half(x):
+    """Rotates half the hidden dims of the input."""
+    x1 = x[..., : x.shape[-1] // 2]
+    x2 = x[..., x.shape[-1] // 2 :]
+    return mx.concatenate([-x2, x1], axis=-1)
+
+def apply_rotary_pos_emb(q, k, cos, sin, unsqueeze_dim=1):
+    """Applies Rotary Position Embedding to the query and key tensors.
+
+    Args:
+        q : The query tensor.
+        k : The key tensor.
+        cos : The cosine part of the rotary embedding.
+        sin : The sine part of the rotary embedding.
+        unsqueeze_dim (`int`, *optional*, defaults to 1):
+            The 'unsqueeze_dim' argument specifies the dimension along which to unsqueeze cos[position_ids] and
+            sin[position_ids] so that they can be properly broadcasted to the dimensions of q and k. For example, note
+            that cos[position_ids] and sin[position_ids] have the shape [batch_size, seq_len, head_dim]. Then, if q and
+            k have the shape [batch_size, heads, seq_len, head_dim], then setting unsqueeze_dim=1 makes
+            cos[position_ids] and sin[position_ids] broadcastable to the shapes of q and k. Similarly, if q and k have
+            the shape [batch_size, seq_len, heads, head_dim], then set unsqueeze_dim=2.
+    Returns:
+        `tuple` comprising of the query and key tensors rotated using the Rotary Position Embedding.
+    """
+    cos = mx.expand_dims(cos, axis=unsqueeze_dim)
+    sin = mx.expand_dims(sin, axis=unsqueeze_dim)
+    q_embed = (q * cos) + (rotate_half(q) * sin)
+    k_embed = (k * cos) + (rotate_half(k) * sin)
+    return q_embed, k_embed
```
mlx.nn already has RoPE

so this is not needed.

Blaizzy replied:
Example: https://github.com/Blaizzy/mlx-vlm/blob/main/mlx_vlm/models/llava/language.py#L78-L83

pappitti replied:
I now remember why I used this instead of the built-in layer. It had nothing to do with the model. I changed to nn.RoPE

Blaizzy commented on mlx_embeddings/models/xlm_roberta.py:
```diff
@@ -339,9 +342,92 @@ def __call__(
         )
-        return (sequence_output, pooled_output) + encoder_outputs[1:]
+
+        if not return_dict:
+            return (sequence_output, pooled_output) + encoder_outputs[1:]
+        
+        return {
+            "embeddings": pooled_output,
+            "last_hidden_state": sequence_output,
+            "hidden_states": encoder_outputs[1] if output_hidden_states else None,
+            "attentions": encoder_outputs[2] if output_attentions else None,
+        }
```
Check if not return dict vs your new return.

Also. pooled_output can be none.

Blaizzy commented:
After looking at sentence transformers files:

- https://huggingface.co/nomic-ai/modernbert-embed-base/blob/main/config_sentence_transformers.json
- https://huggingface.co/Snowflake/snowflake-arctic-embed-l-v2.0/blob/main/config_sentence_transformers.json

I don't think we need them. 

This similarity pipeline is great enough abstraction!

Blaizzy commented on mlx_embeddings/models/xlm_roberta.py:
```diff
@@ -339,9 +342,92 @@ def __call__(
+            return (sequence_output, pooled_output) + encoder_outputs[1:]
+        
+        return {
+            "embeddings": pooled_output,
```
This fixes the embeddings ✅
```suggestion
            "embeddings": sequence_output,
```

Blaizzy replied:
You can add an extra field for pooled_output.

pappitti replied:
I believe that this comment was caused by an error in the ordering for the initial PR. However it highlighted that there may be a confusion about what the models actually return. I tried to clarify but please read my comment on pooling below

pappitti commented:
Thanks for your feedback. It probably goes without saying but I'm not used to doing this. And that's an euphemism.

I believe that I have addressed all the points. and I ran pre-commit checks successfully.  
I add below some comments 

## modernbert.py
NOTE :  removed all the attention_outputs (eager mode), you may want to add it back later given no flash attention 2. padded/unpadded was also removed
TODO: 
- review the padding strategy
- review compiling opportunities

All the training features have been removed

## utils.py
If the HF repo includes a config_sentence_transformers.json file , the pipeline is automatically switched to sentence-similarity.  
There is also another pipeline non-sentence-transformers allowing to run non sentence-transformers models (e.g the original answerai weights) into the sentence-similarity pipeline

## pipelines
Removed the sentence-transformers pipeline, but had to add a non-sentence-transformers pipeline to support non-sentence-transformers specifically for modernBert models to automatically switch to a different weight sanitization strategy. It’s a edge case, and may not be needed

## pooling
bert and xlm-bert pool by using the CLS token, a linear layer and an activation function. However, in my implementation of modernbert, the pooling strategy is defined by a config parameter: classifier_pooling which defaults to ‘mean’ (there is no other parameter in the sentence transformers variants and all of them have 'mean' there). There is effectively no pooler in the modernbert implementation so 'mean' seems the safest way to do it (otherwise it just takes the CLS token). 

## tested models 
- sentence-transformers/all-MiniLM-L6-v2 (bert / sentence-transformers) : single embedding and batch + comparison script
- answerdotai/ModernBERT-base (modernbert) : single embedding and batch + comparison script
- nomic-ai/modernbert-embed-base (modernbert / sentence-transformers) : single embedding and batch + comparison script
- Snowflake/snowflake-arctic-embed-l-v2.0 (xlm-roberta / sentence-transformers) : single embedding and batch + comparison script
- BAAI/bge-small-en-v1.5 (bert / sentence-transformers): single embedding and batch + comparison script
- tasksource/ModernBERT-base-embed  (modernbert / sentence-transformers): single embedding and batch + comparison script

## XLM-RoBERTa
I could only make it work with Snowflake/snowflake-arctic-embed-l-v2.0. For others that I have tested, the problem comes from position_ids parameters in XLMRobertaEmbeddings but I did not look into it too much

Blaizzy commented:
Thanks a lot @pappitti!

Sorry for the delay, I have been swamped.

Will do my best to merge PR as soon as possible.

pappitti commented:
no worries, I am more excited about your mlx-lm and mlx-vlm projects than about this one. I'll probably update the PR at one point to add a zero-shot classification pipeline reflecting this (https://arxiv.org/abs/2502.03793) as it works well on my end

Blaizzy commented:
Awesome! 

@pappitti can you resolve the conflicts and we can merge.

I will make a new release soon

pappitti commented:
apologies, I thought I had fetched upstream but I hadn't.

I resolved the conflicts, it should not impact the siglip workflows but I have not tested extensively

I also changed the README example fro similarities as what was written there did really work.

I tested with these models (2 bert, 3 modernbert, 2 xlm-roberta - mix of sentence transformers and non-sentence transformers modeles)
tested_models = [
    "nomic-ai/modernbert-embed-base",
    "tasksource/ModernBERT-base-embed",
    "answerdotai/ModernBERT-base",
    "sentence-transformers/all-MiniLM-L6-v2",
    "T-Systems-onsite/cross-en-de-roberta-sentence-transformer",
    "Snowflake/snowflake-arctic-embed-l-v2.0",
    "BAAI/bge-small-en-v1.5"
]

ran the pre-commit tests and passed
![Frame 33](https://github.com/user-attachments/assets/b6ccd641-0257-484d-a60a-f500f1b4d6be)

Blaizzy commented:
Awesome! 

Are the XLMRoberta models working for you now? 

We had a recent fix.

pappitti commented:
Yes, as far as i could tell (I tested 2 different models)

Blaizzy commented:
Please check because we added mean pooling and normalisation in #8

Perhaps we don't need a pipeline anymore, but instead just more outputs like SigLip has (including raw and normalized)

Blaizzy commented:
Btw tests are failing could you take a look at it?

pappitti commented:
Oh, ok I understand now where this was coming from. I was a little confused about this.
This was actually a question I had when I first worked on this but I didn't want to touch the bert and xlm-roberta models more than I already had. The pooling strategy and normalization was already covered in the ModernBERT model but not in the others (relying on BertPooler for bert models and XLMRobertaPooler for xlm-roberta models). For my own work I adapted the BertPooler to do the mean pooling and normalization there. But it is equally valid to just get the embeddings and handle the pooling and normalization at the end. I suppose that it is your choice to make. 
So yes, you can absolutely get rid of the pipelines and handle this separately. The Model class returns both pooled outputs and full embeddings anyway (pooling based on Pooler for bert and xlm-roberta and based on config file for modernbert). The only edge case to take into account for modernBert is that the weights for sentence-transformers models and the weights for non-sentence transformers models must be sanitize differently. 

Note that you can do batch processing out of the box with _tokenizer instead of the list comprehension but I've not run benchmarks to check if there was a major benefit
```
input_ids = tokenizer._tokenizer(
    texts, 
    return_tensors="mlx", 
    padding=True, 
    truncation=True, 
    max_length= max_position_embeddings
)
```

pappitti commented:
and yes, I know what the problem is with the tests (models return a dict by default but the tests expects an array).

pappitti commented:
I can work on removing the pipelines tomorrow

Blaizzy commented:
> I can work on removing the pipelines tomorrow

Great looking forward to it :)

pappitti commented:
Hey,

I removed the pipelines and aligned the outputs format for all models with the xlm-roberta model, which I thought was the cleanest of all three. Bert and modernbert now use the mean pooling and normalization functions from the base file for consistency

The readme examples work for all three models (tested with 7 endpoints across the 3 model families)

Also updated the model tests to reflect new output format

Blaizzy commented:
Could you run
> pre-commit run --all

To clear the tests.

pappitti commented:
No, thank YOU for this project. It a great project, I'm excited to test the vision encoders now !

Blaizzy commented:
Could you share the models you tested including ther Sentence Transformers models.

I will fix the loading :)

Blaizzy commented:
> No, thank YOU for this project. It a great project, I'm excited to test the vision encoders now !

My pleasure!

Looking forward to collaborating more with you on future ports

pappitti commented:
here is the full list:

tested_models = [
    "nomic-ai/modernbert-embed-base",
    "tasksource/ModernBERT-base-embed",
    "answerdotai/ModernBERT-base",
    "sentence-transformers/all-MiniLM-L6-v2",
    "T-Systems-onsite/cross-en-de-roberta-sentence-transformer",
    "Snowflake/snowflake-arctic-embed-l-v2.0",
    "BAAI/bge-small-en-v1.5"
]

Blaizzy commented:
Thanks, this helps!
